client/src/connection_cache.rs

```diff
@@ -0,0 +1,155 @@
+use {
+    crate::{tpu_connection::TpuConnection, udp_client::UdpTpuConnection},
+    lazy_static::lazy_static,
+    std::{
+        collections::{hash_map::Entry, BTreeMap, HashMap},
+        net::{SocketAddr, UdpSocket},
+        sync::{Arc, Mutex},
+    },
+};
+
+// Should be non-zero
+static MAX_CONNECTIONS: usize = 64;
+
+struct ConnMap {
+    // Keeps track of the connection associated with an addr and the last time it was used
+    map: HashMap<SocketAddr, (Arc<dyn TpuConnection + 'static + Sync + Send>, u64)>,
+    // Helps to find the least recently used connection. The search and inserts are O(log(n))
+    // but since we're bounding the size of the collections, this should be constant
+    // (and hopefully negligible) time. In theory, we can do this in constant time
+    // with a queue implemented as a doubly-linked list (and all the
+    // HashMap entries holding a "pointer" to the corresponding linked-list node),
+    // so we can push, pop and bump a used connection back to the end of the queue in O(1) time, but
+    // that seems non-"Rust-y" and low bang/buck. This is still pretty terrible though...
+    last_used_times: BTreeMap<u64, SocketAddr>,
+    ticks: u64,
+}
+
+impl ConnMap {
+    pub fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+            last_used_times: BTreeMap::new(),
+            ticks: 0,
+        }
+    }
+}
+
+lazy_static! {
+    static ref CONNECTION_MAP: Mutex<ConnMap> = Mutex::new(ConnMap::new());
+}
+
+#[allow(dead_code)]
+// TODO: see https://github.com/solana-labs/solana/issues/23661
+// remove lazy_static and optimize and refactor this
+pub fn get_connection(addr: &SocketAddr) -> Arc<dyn TpuConnection + 'static + Sync + Send> {
+    let mut map = (*CONNECTION_MAP).lock().unwrap();
+    let ticks = map.ticks;
+
+    let (conn, target_ticks) = match map.map.entry(*addr) {
+        Entry::Occupied(mut entry) => {
+            let mut pair = entry.get_mut();
+            let old_ticks = pair.1;
+            pair.1 = ticks;
+            (pair.0.clone(), old_ticks)
+        }
+        Entry::Vacant(entry) => {
+            let send_socket = UdpSocket::bind("0.0.0.0:0").unwrap();
```

Comment:
Can you bind this multiple times concurrently for different addresses?

Reply:
It's the same as binding to `INADDR_ANY`

Reply:
okay -- earlier this was only bound once -- and the same send socket is used for all different addresses. Now we are create that for each unique address. That is a net change. Not sure if we need to worry about it.